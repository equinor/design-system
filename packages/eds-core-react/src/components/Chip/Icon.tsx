import styled, { css } from 'styled-components'
import { close } from '@equinor/eds-icons'
import { Icon as Icon_ } from '../Icon'
import * as tokens from './Chip.tokens'
import { bordersTemplate } from '@equinor/eds-utils'

Icon_.add({ close })

const { enabled, error } = tokens

type IconProps = {
  $variant: 'active' | 'error' | 'default'
  $disabled: boolean
}

export const Icon = styled(Icon_)<IconProps>`
  cursor: pointer;
  padding: 1px;
  ${bordersTemplate(enabled.entities.icon.border)}
  &:hover {
    ${({ $variant }) => {
      switch ($variant) {
        case 'error':
          return css`
            background: ${error.entities.icon.background};
            svg {
              fill: ${error.entities.icon.typography.color};
            }
          `
        default:
          return css`
            background: ${enabled.entities.icon.states.hover.background};
            svg {
              fill: ${enabled.states.hover.typography.color};
            }
          `
      }
    }};
  }

  ${({ $disabled }) =>
    $disabled &&
    css`
      cursor: not-allowed;
      &:hover {
        background: transparent;
      }
    `}
`
