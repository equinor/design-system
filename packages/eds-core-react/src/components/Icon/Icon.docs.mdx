
import { Links, Story, PropsTable } from './../../../.storybook/components'
import * as icons from '@equinor/eds-icons'
import { Icon  } from '.'


# Icon 
Component for rendering icons from the `@equinor/eds-icons` package.

<Links 
    figmaUrl="https://www.figma.com/file/BQjYMxdSdgRkdhKTDDU7L4KU/Assets?node-id=2%3A3"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/347dd6-system-icons/b/25520a"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Icon/Icon.tsx"
/>

<Story id="icons-icon--introduction" />
<PropsTable />

## Installation 

Install the packages in your project directory with your package manager (npm/yarn/pnpm).
```shell
npm install @equinor/eds-core-react @equinor/eds-icons
```

## Usage

The `Icon` can then render the icons in two ways: 

### Option 1 

Using the `name` property. This allows you to setup all your icons once and use the `name` property to render it.
```tsx
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons' // import "save" icon

Icon.add({ save }) // needs only to be done once

<Icon name="save" />
```
### Option 2

Use the `data` property for directly rendering the imported icon.
```tsx
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons' // import "save" icon

<Icon data={save} />
```
**Checkout the [EDS library](?path=/docs/icons--preview) for available icons**

## Accessibility
If you are using the icons alone to convey meaning you need to provide the `title` property so that your icons are correctly accessible.
```tsx
 <Icon name="save" title="Save me!" />
```
<Story id="icons-icon--accessibility" />

## Examples

### Colour
<Story id="icons-icon--color" />


### Rotations
<Story id="icons-icon--rotations" />


### Sizes
Simple and clear icons require a strict underlying grid.
<ul>
    <li>When using compact mode the size of icons are 18px.</li>
</ul>
<Story id="icons-icon--sizes" />





