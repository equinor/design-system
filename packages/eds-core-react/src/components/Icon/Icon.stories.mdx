import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs'
import styled from 'styled-components'
import { Icon  } from '.'
import * as stories from './Icon.stories.tsx'
import * as icons from '@equinor/eds-icons'


<Meta 
  title="Icons/Icon"
  component={Icon}
  args={{name:"save"}}
  argTypes={{
    name: {
      options: Object.keys(icons),
      control: {
        type: 'select',
      },
    },
    color: {
      control: 'color',
    },
  }}
  />

export const Abbr = styled.abbr`
  cursor: help;
`

# Icon 
Component for rendering icons from the `@equinor/eds-icons` package.

## Installation 

Install the packages in your project directory with your npm package manager
```shell
npm install @equinor/eds-core-react @equinor/eds-icons
```

## Usage

The `Icon` can then render the icons in two ways: 

### Option 1 

Using the `name` property. This allows you to setup all your icons once and use the `name` property to render it.
```tsx
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons' // import "save" icon

Icon.add({ save }) // needs only to be done once

<Icon name="save" />
```
### Option 2

Use the `data` property for directly rendering the imported icon
```tsx
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons' // import "save" icon

<Icon data={save} />
```

## Accessibility
If you are using the icons semantically you need to provide the `title` property so that your icons are correctly accessible.
```tsx
 <Icon name="save" title="Save me!" />
```


## Color
<Canvas>
  <Story story={stories.Color} />
</Canvas>



## Rotations
<Canvas>
  <Story story={stories.Rotations} />
</Canvas>


## Sizes
<Canvas>
  <Story story={stories.Sizes} />
</Canvas>


## Props
<Canvas>
  <Story story={stories.Default} />
</Canvas>

<ArgsTable of={Icon} />


