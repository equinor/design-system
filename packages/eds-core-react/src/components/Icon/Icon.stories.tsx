import { Story, ComponentMeta } from '@storybook/react'
import * as icons from '@equinor/eds-icons'
import { Stack } from './../../../.storybook/components'
import { Icon, IconProps } from '../..'
import page from './Icon.docs.mdx'

Icon.add(icons)

export default {
  title: 'Icons/Icon',
  component: Icon,
  argTypes: {
    name: {
      options: Object.keys(icons),
      control: {
        type: 'select',
      },
    },
    color: {
      control: 'color',
    },
  },
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Icon>

export const Introduction: Story<IconProps> = (args) => (
  <Stack>
    <Icon {...args} />
  </Stack>
)

Introduction.args = {
  name: 'save',
  size: 24,
  title: 'save',
}

export const Accessibility: Story<IconProps> = () => (
  <Stack>
    <Icon name="save" title="Save me!" />
  </Stack>
)

export const Color: Story<IconProps> = () => (
  <Stack>
    <Icon name="save" color={'red'} />
    <Icon name="save" color={'green'} />
    <Icon name="save" color={'blue'} />
  </Stack>
)

export const Rotations: Story<IconProps> = () => (
  <Stack>
    <Icon name="pregnant_woman" rotation={90} />
    <Icon name="pregnant_woman" rotation={180} />
    <Icon name="pregnant_woman" rotation={270} />
  </Stack>
)

export const Sizes: Story<IconProps> = () => (
  <Stack>
    <Icon name="fullscreen" size={16} />
    <Icon name="fullscreen" size={18} />
    <Icon name="fullscreen" size={24} />
    <Icon name="fullscreen" size={32} />
    <Icon name="fullscreen" size={40} />
    <Icon name="fullscreen" size={48} />
  </Stack>
)
