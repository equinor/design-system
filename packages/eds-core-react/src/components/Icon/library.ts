import type { IconData } from '@equinor/eds-icons'
import type { Name } from './Icon.types'

type IconRecord = Record<Name, IconData>

let _icons: IconRecord = {}
/** Add icons to library to be used for rendering using name.
This needs to be done lonly once */
export const add = (icons: IconRecord): void => {
  _icons = {
    ..._icons,
    ...icons,
  }
}

export const get = (name: Name): IconData | undefined => {
  return _icons[name]
}
