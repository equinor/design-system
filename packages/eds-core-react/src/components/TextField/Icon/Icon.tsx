import { HTMLAttributes, forwardRef } from 'react'
import styled, { css } from 'styled-components'
import { useTextField } from '../TextField.context'
import { input as tokens } from './Icon.tokens'
import type { Variants, ColorStateProps } from '../types'

type StyledIconProps = {
  colors: ColorStateProps
  isDisabled?: boolean
  isFocused?: boolean
  size?: 16 | 24
}

const StyledIcon = styled.div<StyledIconProps>(
  ({ colors, isDisabled, isFocused, size }) => {
    const { focusColor, color, disabledColor } = colors
    let fill = color

    if (isDisabled) {
      fill = disabledColor
    }
    if (isFocused) {
      fill = focusColor
    }

    return css`
      &,
      svg {
        fill: ${fill};
        width: ${size}px;
        height: ${size}px;
      }
    `
  },
)

type TextfieldIconProps = {
  /** isDisabled */
  isDisabled?: boolean
  /** Variant */
  variant?: Variants
  /** Colors */
  colors?: ColorStateProps
  /** Size */
  size?: 16 | 24
} & HTMLAttributes<HTMLDivElement>

const InputIcon = forwardRef<HTMLDivElement, TextfieldIconProps>(
  function InputIcon(
    {
      size = 24,
      variant = 'default',
      isDisabled = false,
      colors = {
        color: tokens[variant].color,
        disabledColor: tokens[variant].disabledColor,
        focusColor: tokens[variant].focusColor,
      },
      children,
      ...other
    },
    ref,
  ) {
    const { isFocused } = useTextField()

    const iconProps = {
      isDisabled,
      colors,
      isFocused,
      size,
    }

    return (
      <StyledIcon ref={ref} {...iconProps} {...other}>
        {children}
      </StyledIcon>
    )
  },
)

export { InputIcon as Icon }
