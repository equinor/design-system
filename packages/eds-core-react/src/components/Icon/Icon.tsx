import { forwardRef, Ref, SVGProps } from 'react'
import type { IconData } from '@equinor/eds-icons'
import styled from 'styled-components'
import type { Name } from './Icon.types'
import { get } from './library'

type StyledProps = {
  height: number
  $height?: number
  width: number
  fill: string
  size?: number
  rotation?: number
}

type SimpleSVGProps = {
  name: string
  viewBox: string
  rotation?: number
  title?: string
  role?: string
  'aria-hidden'?: boolean | 'true' | 'false'
  'aria-labelledby'?: string
}

const transform = ({ rotation }: SimpleSVGProps): string =>
  rotation ? `transform: rotate(${rotation}deg)` : ''

const StyledSvg = styled.svg.attrs<StyledProps>(({ height, width, fill }) => ({
  name: null,
  xmlns: 'http://www.w3.org/2000/svg',
  height: `${height}px`,
  width: `${width}px`,
  fill,
}))`
  ${transform}
`

const StyledPath = styled.path.attrs<StyledProps>(({ $height, size }) => ({
  size: null,
  fillRule: 'evenodd',
  clipRule: 'evenodd',
  transform: size / $height !== 1 ? `scale(${size / $height})` : null,
}))``

const customIcon = (icon: IconData) => ({
  icon,
  count: Math.floor(Math.random() * 1000),
})

const findIcon = (name: string, data: IconData, size: number) => {
  // eslint-disable-next-line prefer-const
  let { icon, count } = data ? customIcon(data) : get(name)

  if (size < 24) {
    // fallback to normal icon if small is not made yet
    icon = icon.sizes?.small || icon
  }

  return { icon, count }
}

export type IconProps = {
  /** Title for icon when used semantically */
  title?: string
  /** Color */
  color?: string
  /** Size */
  size?: 16 | 18 | 24 | 32 | 40 | 48
  /** Rotation */
  rotation?: 0 | 90 | 180 | 270
  /** Name */
  name?: Name
  /** Manually specify which icon data to use */
  data?: IconData
  /** @ignore */
  ref?: Ref<SVGSVGElement>
} & SVGProps<SVGSVGElement>

export const Icon = forwardRef<SVGSVGElement, IconProps>(function Icon(
  { size, color = 'currentColor', name, rotation, title, data, ...rest },
  ref,
) {
  // eslint-disable-next-line prefer-const
  const { icon, count } = findIcon(name, data, size)

  if (typeof icon === 'undefined') {
    throw Error(
      `Icon "${name}" not found. Have you added it using Icon.add() or using data props?`,
    )
  }

  const height = size ? size : parseInt(icon.width)
  const width = size ? size : parseInt(icon.height)

  let svgProps: SimpleSVGProps & StyledProps = {
    height,
    width,
    fill: color,
    viewBox: `0 0 ${width} ${height}`,
    rotation,
    name,
    'aria-hidden': true,
  }

  const pathProps = {
    d: icon.svgPathData,
    $height: icon.height ? icon.height : size,
    size: size || icon.height,
  }

  // Accessibility
  let titleId = ''

  if (title) {
    titleId = `${icon.prefix}-${icon.name}-${count}`
    svgProps = {
      ...svgProps,
      title,
      role: 'img',
      'aria-hidden': undefined,
      'aria-labelledby': titleId,
    }
  }

  return (
    <StyledSvg {...svgProps} {...rest} ref={ref}>
      {title && <title id={titleId}>{title}</title>}
      <StyledPath data-testid="eds-icon-path" {...pathProps} />
    </StyledSvg>
  )
})

// Icon.displayName = 'EdsIcon'
