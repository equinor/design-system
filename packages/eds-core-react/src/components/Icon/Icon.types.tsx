import type { IconData, IconName } from '@equinor/eds-icons'
import { Icon } from './Icon'
import { add } from './library'

export type IconBasket = { icon?: IconData; count: number }

export type Name = IconName

export type IconType = typeof Icon & {
  add: typeof add
}
