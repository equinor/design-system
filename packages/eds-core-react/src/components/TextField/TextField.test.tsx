/* eslint-disable no-undef */
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import 'jest-styled-components'
import styled from 'styled-components'
import { TextField } from '.'
import { Icon } from '../Icon'
import { save } from '@equinor/eds-icons'

afterEach(cleanup)

describe('TextField', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <TextField id="textfield" label="textfield" />,
    )

    expect(asFragment).toMatchSnapshot()
  })
  it('Has correct label text', () => {
    const labelText = 'Some label'
    render(<TextField id="test-label" label={labelText} />)

    expect(screen.queryByText(labelText)).toBeInTheDocument()
  })

  it('Has correct default value', () => {
    const value = 'Some value'
    render(<TextField id="test-value" value={value} readOnly />)
    const inputElement: HTMLInputElement = screen.queryByDisplayValue(value)

    expect(inputElement.value).toBe(value)
  })

  it('Has correct updated value when changed', () => {
    let value = 'Initial value'
    const newValue = 'Updated value'
    const onChangeHandler = (ele) => {
      // eslint-disable-next-line
      value = ele.target.value
    }

    render(
      <TextField
        id="test-value-updated"
        value={value}
        onChange={onChangeHandler}
      />,
    )

    fireEvent.change(screen.getByDisplayValue(value), {
      target: { value: newValue },
    })

    expect(value).toBe(newValue)
  })

  it('Can have a helper text', () => {
    const helperText = 'Help me'

    render(<TextField id="test" helperText={helperText} />)
    const inputNode = screen.getByText(helperText)
    expect(inputNode).toBeDefined()
  })

  it('Can have an icon inside the input', () => {
    const labelText = 'Label for test'
    render(
      <TextField
        id="test"
        label={labelText}
        inputIcon={<Icon data={save} />}
      />,
    )
    // eslint-disable-next-line testing-library/no-node-access
    const inputNode = screen.getByLabelText(labelText).parentElement
    // eslint-disable-next-line testing-library/no-node-access
    expect(inputNode.querySelectorAll('svg')).toHaveLength(1)
  })
  it('Can have an icon by the helper text', () => {
    const helperText = 'Help me'
    render(
      <TextField
        id="test"
        label="hi"
        helperText={helperText}
        helperIcon={<Icon data={save} />}
      />,
    )
    // eslint-disable-next-line testing-library/no-node-access
    const helperNode = screen.getByText(helperText).parentNode
    // eslint-disable-next-line testing-library/no-node-access
    expect(helperNode.querySelectorAll('svg')).toHaveLength(1)
  })

  const StyledTextField = styled(TextField)`
    margin-top: 48px;
  `
  it('Can extend the css of the component', () => {
    const { container } = render(
      <StyledTextField id="test-css-extend" variant="error" />,
    )

    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveStyleRule('margin-top', '48px')
  })
})
