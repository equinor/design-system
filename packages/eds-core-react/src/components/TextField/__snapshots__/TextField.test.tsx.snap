// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Matches snapshot 1`] = `
<DocumentFragment>
  .c4 {
  width: 100%;
  box-sizing: border-box;
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--eds_ui_background__light,rgba(247,247,247,1));
  border: none;
  height: 36px;
  box-shadow: inset 0px -1px 0px 0px var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
  outline: 1px solid transparent;
  outline-offset: 0px;
  margin: 0;
  color: var(--eds_text__static_icons__default,rgba(61,61,61,1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.500em;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  text-align: left;
  padding-left: 8px;
  padding-top: 6px;
  padding-right: 8px;
  padding-bottom: 6px;
}

.c4::-webkit-input-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c4::-moz-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c4:-ms-input-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c4::placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c4:active,
.c4:focus {
  outline-offset: 0;
  box-shadow: none;
  outline: 2px solid var(--eds_interactive_primary__resting,rgba(0,112,121,1));
  outline-offset: 0px;
}

.c4:disabled {
  color: var(--eds_interactive__disabled__text,rgba(190,190,190,1));
  cursor: not-allowed;
  box-shadow: none;
  outline: none;
}

.c4:disabled:focus,
.c4:disabled:active {
  outline: none;
}

.c4[readOnly] {
  background: transparent;
  box-shadow: none;
}

.c5 {
  outline: none;
}

.c5:active,
.c5:focus {
  outline: none;
  box-shadow: none;
}

.c3 {
  box-shadow: inset 0 -1px 0 0 #6f6f6f;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  margin: 0;
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
  font-family: Equinor;
  font-size: 0.750rem;
  font-weight: 500;
  line-height: 1.333em;
  text-align: left;
  margin-left: 8px;
  margin-right: 8px;
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c2 {
  margin: 0;
}

.c0 {
  min-width: 100px;
  width: 100%;
}

<div
    class="c0"
  >
    <label
      class="c1"
      for="textfield"
    >
      <span
        class="c2"
      >
        textfield
      </span>
    </label>
    <div
      class="c3"
    >
      <input
        aria-describedby="eds-helpertext-12346"
        aria-invalid="false"
        class="c4 c5"
        id="textfield"
        type="text"
      />
    </div>
  </div>
</DocumentFragment>
`;
